src/lib.rs

```diff
@@ -442,21 +473,88 @@ impl Log for Logger {
     }
 
     fn log(&self, record: &Record) {
-        let _ = match self.target {
-            Target::Stdout if self.matches(record) => (self.format)(&mut io::stdout(), record),
-            Target::Stderr if self.matches(record) => (self.format)(&mut io::stderr(), record),
-            _ => Ok(()),
-        };
+        if self.matches(record) {
+            // Log records are written to a thread-local buffer before being printed
+            // to the terminal. We clear these buffers afterwards, but they aren't shrinked
+            // so will always at least have capacity for the largest log record formatted
+            // on that thread.
+
+            thread_local! {
+                static FORMATTER: RefCell<Option<Formatter>> = RefCell::new(None);
+            }
+
+            FORMATTER.with(|tl_buf| {
+                let mut tl_buf = tl_buf.borrow_mut();
+
+                if tl_buf.is_none() {
+                    *tl_buf = Some(Formatter::new(self.writer.buffer()));
+                }
+
+                // The format is guaranteed to be `Some` by this point
+                let mut formatter = tl_buf.as_mut().unwrap();
+
+                let _ = (self.format)(&mut formatter, record)
```

Comment:
I think this should clarify its error conditions a bit better:

- We only print the record to the terminal if the record is successfully formatted (which it pretty much always should)
- We always clear the buffer even if the record wasn't printed. Seems like it might not be ideal, but I'm not really sure what would cause a print to fail.

I think it needs a little more thought and should be refactored to clarify the intent a bit better.